### ISSUE TITLE
ColorPicker for ObjectProperty<Color>

### ISSUE THREAD

**Issue by gluon-bot** [gluonhq/scenebuilder#84]
Originally reported by: **Gerrit Grunwald (Bitbucket: [hansolo](https://bitbucket.org/hansolo), GitHub: [hansolo](https://github.com/hansolo))**

----------------------------------------

It seems that properties in controls of type ObjectProperty<Paint> can be edited in the SceneBuilder UI but properties of type ObjectProperty<Color> can't.
So properties of type ObjectProperty<Color> just appear as disabled textfields. It would be great if one could edit those properties by using a ColorPicker instead of the combined Color/Paint picker.

----------------------------------------
- Bitbucket: https://bitbucket.org/gluon-oss/scenebuilder/issue/84

**Comment by abhinayagarwal** [comment]
Hi @hansolo,

**tl;dr** Scene Builder cannot show a ColorPicker instead of disabled TextField because of the Type Erasure limitation in Java. 

**Explanation**

Scene Builder iterates through all properties of a control and depending on the property type it creates a control. This is all handled in [InspectorPanelController#makeOrResetPropertyEditor](https://github.com/gluonhq/scenebuilder/blob/master/kit/src/main/java/com/oracle/javafx/scenebuilder/kit/editor/panel/inspector/InspectorPanelController.java#L1757).

`ColorPicker`, `DatePicker` etc. extends from `ComboBoxBase<T>`. The property `value` is of type `ObjectProperty<T>`. At runtime, its impossible to find the type of `T` and therefore we cannot create a control from the same. Therefore, Scene Builder default to [GenericEditor](https://github.com/gluonhq/scenebuilder/blob/master/kit/src/main/java/com/oracle/javafx/scenebuilder/kit/editor/panel/inspector/editors/GenericEditor.java) which is a disabled TextField.

If someone knows how to by-pass this limitation, it would be a great addition to Scene Builder :)

**Comment by mimmoz81** [comment]
Well, that's not completely true..
Take a look at this simple test:

	import java.lang.reflect.Field;
	import java.lang.reflect.Type;
	
	import javafx.beans.property.SimpleObjectProperty;
	import javafx.scene.paint.Color;
	
	public class GenericTypeExtraction {
	
		private SimpleObjectProperty<Color> color = new SimpleObjectProperty<>();
		
		public static void main(String[] args) {
			try {
				GenericTypeExtraction g = new GenericTypeExtraction();
				Field field = g.getClass().getDeclaredField("color");
				Type type = field.getGenericType();
				
				String typeName = type.getTypeName();
				int start = typeName.indexOf("<");
				int end = typeName.indexOf(">");
				if (start >= 0 && end > start)
					System.out.println(typeName.substring(start + 1, end)); // this prints "javafx.scene.paint.Color"
				else
					System.out.println("No generic type found");
				
			} catch (Exception e) {
				e.printStackTrace();
			}
			
		}
	}

With some NPE-safe-tests, this procedure can be used to extract the class from the generic type of the property. A Class.forName(..) will do the rest eventually..

**Comment by abhinayagarwal** [comment]
With further investigation, I have found that this is indeed possible. But, there is a catch.

Currently in SB, `ColorPropertyMetadata` returns a hex string of a color. This hex string is being used by classes which has `color` property in them e.g. `DropShadow`. 

```
<DropShadow color="#ffee00"/>
```

`BeanAdapter` in `FXMLLoader` could easily interpret the above line by looking into the [setter](https://github.com/javafxports/openjdk-jfx/blob/4f0addff16de7182752bc60541f44baeb0e99d2a/modules/javafx.fxml/src/main/java/com/sun/javafx/fxml/BeanAdapter.java#L250) and [getter](https://github.com/javafxports/openjdk-jfx/blob/4f0addff16de7182752bc60541f44baeb0e99d2a/modules/javafx.fxml/src/main/java/com/sun/javafx/fxml/BeanAdapter.java#L337) methods of key "color". This lead to reflective initializing of `Color` class and passing the value to `valueOf()` since Color doesn't have a default constructor. 

In case of `ObjectProperty<T>`, the setter/getter methods are inferred as type Object and therefore `Color#valueOf` is never called. This is something which needs to be fixed in `FXMLLoader`.

Although, there is an easy way to overcome this in Scene Builder.  We can update `ColorPropertyMetadata` to return a more informative result and `FXMLLoader` can use it to construct a Color object using its [constructor](https://docs.oracle.com/javase/8/javafx/api/javafx/scene/paint/Color.html#Color-double-double-double-double-).

I propose to update:

```
<DropShadow color="#ffee00"/>
```

to 

```
<DropShadow>
    <color>
        <Color blue="0.0972222238779068" green="0.7064814567565918" red="0.75" />
    </color>
</DropShadow>
```

This is backward compatible with the earlier FXML. SB can open the former but will save the FXML file as latter.

This can be then used for ColorPicker as well:

```
<ColorPicker>
    <value>
        <Color blue="0.0972222238779068" green="0.7064814567565918" red="0.75" />
    </value>
</ColorPicker>
```

**Comment by abhinayagarwal** [comment]
ColorPicker is not the optimal control for choosing color because the custom color dialog leaves the bounds of the screen when Scene Builder is in full screen mode. As of today, there is no API to change the side on which the custom color appears on the ColorPicker popup.

I have therefore [fallen back to a PaintPicker](https://github.com/gluonhq/scenebuilder/pull/209) which only shows the SB's ColorPicker for choosing the value of the JavaFX ColorPicker control.

With #209, one can edit and change the value of JavaFX ColorPicker via Scene Builder.
